### DISCUSSION TITLE
Deleting data

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#1586] [Ideas]
At present, Medplum supports the following methods for deleting data:
1. The FHIR `DELETE` operation, which performs a soft delete
2. Project membership "delete", which deprovisions a user account
3. SCIM account "delete", which deprovisions a user account
4. Super admin "purge" operation to permanently delete data per legal requirements

These are good starting primitives, because it provides the building blocks of functionality to fulfill most requirements.

However, there are limitations / opportunities for improvement.

1. Delete patient record - delete all data for a patient
2. Delete on deprovision - when a user account is deleted, also delete all related data
3. Expose "purge" to project administrators

## References

Smile CDR - Deleting Data
https://smilecdr.com/docs/fhir_repository/deleting_data.html

Aidbox - Deleting Data
https://docs.aidbox.app/storage-1/delete-data

Google Healthcare - Resource Purge
https://cloud.google.com/healthcare-api/docs/reference/rest/v1/projects.locations.datasets.fhirStores.fhir/Resource-purge

IBM FHIR - Resource Erase
https://github.com/LinuxForHealth/FHIR/issues/850

**Comment by reshmakh** [comment]
Regarding the improvements, some thoughts:

1. This seems sensible and convenient
2. I don't think we should delete on deprovision.  It will be common that people will cancel their membership - but the practice will still need to keep their data for recordkeeping purposes.  Deprovision should leave resources in place IMO.  Having a way to query the deleted memberships seems more correct IMO.
3. This is potentially useful, but room for abuse/error and seems like gating behind super admin would be aligned with the controls needed in practice.

**Comment by codyebberson** [comment]
This is all implemented.  Docs: https://github.com/medplum/medplum/pull/1846
